util/network/src/client_version.rs

```diff
@@ -0,0 +1,473 @@
+
+use semver::Version;
+use std::fmt;
+
+
+/// Description of the software version running in a peer
+/// according to https://github.com/ethereum/wiki/wiki/Client-Version-Strings
+/// This structure as it is represents the format used by Parity clients. Other
+/// vendors may provide additional fields.
+///
+/// TODO support formats with extra fields, e.g.:
+/// "Geth/main.jnode.network/v1.8.21-stable-9dc5d1a9/linux-amd64/go1.11.4"
+
+#[derive(Clone,Debug,PartialEq,Eq)]
+pub struct ParityClientData {
+	name: String,
+	variant: Option<String>,
+	semver: Version,
+	os: String,
+	compiler: String,
+}
+
+// Accessor methods for ParityClientData. This will probably
+// need to be abstracted away into a trait.
+impl ParityClientData {
+	fn name(&self) -> &str {
+		self.name.as_str()
+	}
+
+	fn variant(&self) -> Option<&str> {
+		self.variant.as_ref().map(String::as_str)
+	}
+
+	fn semver(&self) -> &Version {
+		&self.semver
+	}
+
+	fn os(&self) -> &str {
+		self.os.as_str()
+	}
+
+	fn compiler(&self) -> &str {
+		self.compiler.as_str()
+	}
+}
+
+#[derive(Clone,Debug,PartialEq,Eq)]
+pub enum ClientVersion {
+	ParityClient(
+		ParityClientData
+	),
+	ParityUnknownFormat(String),
+	Other(String), // Id string
+}
+
+// TODO: Maybe merge with Peercapabilityinfo in ethcore-network?
+pub trait ClientCapabilities {
+	fn can_handle_large_requests(&self) -> bool;
+
+	fn accepts_service_transaction(&self) -> bool;
+}
+
+// This is an implementation of a function taken from propagator.rs
+fn parity_accepts_service_transaction(parity_client_data: &ParityClientData) -> bool {
+	let service_transactions_version = Version::parse(SERVICE_TRANSACTIONS_VERSION).unwrap();
+
+	*parity_client_data.semver() >= service_transactions_version
+}
+
+impl ClientCapabilities for ClientVersion {
+	fn can_handle_large_requests(&self) -> bool {
+		match self {
+			ClientVersion::ParityClient(data) => {
+				if *data.semver() < Version::parse(PARITY_CLIENT_LARGE_REQUESTS_VERSION).unwrap() {
+					false
+				} else {
+					true
+				}
+			},
+			ClientVersion::ParityUnknownFormat(_) => false, // Play it safe
+			ClientVersion::Other(_) => true // As far as we know
+		}
+	}
+
+	/// Checks if peer is able to process service transactions
+	fn accepts_service_transaction(&self) -> bool {
+		match self {
+			ClientVersion::ParityClient(data) => parity_accepts_service_transaction(&data),
+			ClientVersion::ParityUnknownFormat(_) => false,
+			ClientVersion::Other(_) => false
+		}
+	}
+
+}
+
+
+fn is_parity(client_id: &str) -> bool {
+	if client_id.starts_with(LEGACY_CLIENT_ID_PREFIX) || client_id.starts_with(PARITY_CLIENT_ID_PREFIX) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+// Parse known parity formats.
+//
+// This is really not robust: parse four arguments and
+// allow for an extra argument between identifier and
+// version
+// TODO implement a better logic
+fn parse_parity_format(client_version: &str) -> Result<ParityClientData, ()> {
```

Comment:
The logic seems a bit tricky. if `variant` starts with `v` we are going to treat it as version and panic.

Why not change it so that we check the lengths first and then based on the fact that it has either 5 or 4 elements do different parsing of components?

Reply:
I'm not happy with it either. The specification of the format is very loose and every client is more or less free to write their ID as they want (https://github.com/ethereum/wiki/wiki/Client-Version-Strings). We could use a regex to match the version part instead of just checking for a prefixing 'v'.

Reply:
@elferdo I'd say to focus on the variants we know exist and we want to support. So afaict we have either:
- 4-parts client version id (version on the second position)
- 5-parts client version id (variant on the second position and version on the third position)

If we can't parse the string as one of the two variants then we can just assume "unknown". Note that the stuff that changes depending on the version is just some optimizations, i.e.:
1. Don't send service transactions to peers that you know will reject them
2. Don't request more headers if you know that the peer will reject them (we can always fine a different peer to request from anyway).

So we don't have to be super strict on parsing all possible cases here.

Reply:
Right, agreed. Let's focus on specific, known instances and only try to generalize when the need comes.